lib/matplotlib/figure.py

```diff
@@ -2033,6 +1993,14 @@ def _do_layout(gs, mosaic, unique_ids, nested):
             # nested mosaic) at this level
             this_level = dict()
 
+            # When dealing with .add_subfigure - need to create
+            # separate SubFigures for nested layouts,
+            # otherwise they will all create without nesting.
+            subfigs_for_nested = {}
+            if sub_add_func.__name__ == 'add_subfigure':
```

Comment:
Checking the name of a passed in method is generally a bad sign!

To understand the problem here it is that when there are nested sub figures we need to switch which (sub) figures `add_subfigure` we pass down through?

Reply:
Essentially, yes. I wanted to keep track of all of the nested figures (in any 'level' of nesting) with the `subfigs_for_nested` dictionary, so that later on I could assign the correct nested figures their parent 'dummy' figure (and they got positioned correctly). Since `subplot_mosaic` worked just fine with nesting already, I did this.